cmd/microcloud/session.go

```diff
@@ -0,0 +1,166 @@
+		fmt.Println("Waiting for systems to send their intent to join ...")
+	}
+
+	confirmedIntents, err := c.askJoinIntents(gw, expectedSystems)
+	if err != nil {
+		return err
+	}
+
+	if !c.autoSetup {
+		cloud := sh.Services[types.MicroCloud].(*service.CloudService)
+		cert, err := cloud.Certificate()
+		if err != nil {
+			return err
+		}
+
+		fingerprint, err := c.shortFingerprint(cert.Fingerprint())
+		if err != nil {
+			return fmt.Errorf("Failed to shorten fingerprint: %w", err)
+		}
+
+		fmt.Printf(" Using fingerprint %q to confirm the join intents\n", fingerprint)
+	}
+
+	err = gw.Write(types.Session{
+		ConfirmedIntents: confirmedIntents,
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to send join intents: %w", err)
+	}
+
+	err = gw.ReceiveCombined(&session)
+	if err != nil {
+		return fmt.Errorf("Failed to read confirmation errors: %w", err)
+	}
+
+	if !session.Accepted {
+		return fmt.Errorf("Join confirmations didn't get accepted on all systems")
+	}
+
+	for _, joinIntent := range confirmedIntents {
+		certBlock, _ := pem.Decode([]byte(joinIntent.Certificate))
+		if certBlock == nil {
+			return fmt.Errorf("Invalid certificate file")
+		}
+
+		remoteCert, err := x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+
+		// Register init system
+		c.systems[joinIntent.Name] = InitSystem{
+			ServerInfo: mdns.ServerInfo{
+				Version:  joinIntent.Version,
+				Name:     joinIntent.Name,
+				Address:  joinIntent.Address,
+				Services: joinIntent.Services,
+				// Store the peers certificate to allow mTLS server validation
+				// for requests after the trust establishment.
+				Certificate: remoteCert,
+			},
+		}
+	}
+
+	// Add a space between the CLI and the response.
+	fmt.Println("")
+
+	if !c.autoSetup {
+		for _, info := range c.systems {
+			fmt.Printf(" Selected %q at %q\n", info.ServerInfo.Name, info.ServerInfo.Address)
+		}
+	}
+
+	// Add a space between the CLI and the response.
+	fmt.Println("")
+
+	return nil
+}
+
+func (c *initConfig) joiningSession(gw *cloudClient.WebsocketGateway, sh *service.Handler, services []types.ServiceType, target string, passphrase string) error {
+	session := types.Session{
+		Passphrase:    passphrase,
+		Address:       sh.Address,
+		TargetAddress: target,
+		Interface:     c.lookupIface.Name,
+		Services:      services,
+		LookupTimeout: c.lookupTimeout,
+	}
+
+	err := gw.Write(session)
+	if err != nil {
+		return fmt.Errorf("Failed to send session start: %w", err)
+	}
+
+	if !c.autoSetup && target == "" {
+		fmt.Println("Searching an eligible system ...")
+	}
+
+	// The server confirms the target address if none was provided by the client.
+	err = gw.ReceiveCombined(&session)
+	if err != nil {
+		return fmt.Errorf("Failed to find an eligible system: %w", err)
+	}
+
+	if !c.autoSetup {
+		fmt.Printf("Waiting for system %q to confirm our intent to join ...\n", session.TargetAddress)
```

Comment:
Is there a way to get the name of the initiator at this stage? On some lines we refer to the system by name, and on this line we do so by address. It would be nice to keep it consistent.

Also, how about changing this so that the next step is clear to the user:
```
Complete cluster member selection on "m1" to continue ...
```

Reply:
Ah I see so in the unicast case this is just the user-specified target address?

Reply:
Yep, the server checks if the target address is set, if not use mDNS to lookup a target, but in any case return the final target back to the client.

Reply:
Reworked as discussed.

First we lookup the address in case of multicast or continue if the unicast address was already provided.
As the mDNS payload was reduced down to only a minimal set of information we query the status endpoint in both the multicast and unicast scenario to retrieve the required information equally in both cases.

We now query the status endpoint on the target to retrieve the remote's name.
The fingerprint is resolved by using the `shared.GetRemoteCertificate` func.